### DISCUSSION TITLE
Managing Multiple Cellars in Homebrew for Flexible Development Environments

### DISCUSSION THREAD

**Discussion by RootRooster** [Homebrew/discussions#5203] [Everyday usage] [answered]
### Output of `brew config`

```shell
.
```

### Output of `brew doctor`

```shell
.
```

### Description of issue

Hello everyone,

As a CS student, I'm looking for ways to keep my Mac organized while working on various coding projects. While I've been using brew for basic tasks, I'm now interested in using Docker to manage different technology stacks efficiently within VSCode development environments.

I'm curious if there's a way to achieve similar flexibility with brew, such as having multiple cellars that users can switch between. For instance, I'd like to have a cellar for school work that includes Java 8, and another cellar for work with Java 15. Switching between these cellars should be easy and not require re-downloading packages. Is this possible with brew?

If not; would this be a feasible feature to develop for brew, and could it benefit other developers? 

Best regards,
N

**Comment by ZhongRuoyu** [comment, marked as answer]
Switching between cellars is not available in Homebrew at the moment. There was a proposal, though it became stale:

- Homebrew/brew#11787

Homebrew does allow versioned formulae to be installed simultaneously (`openjdk`, `python`, `node`, ...), although for most of them only one version can be linked at a time (with `python` being a notable exception). Different shell environments can be configured for individual projects to cater for version preferences. [^1]

Also, since you mentioned you use VS Code: you can configure different Java runtimes for different projects [with a VS Code extension](https://code.visualstudio.com/docs/java/java-project#_configure-runtime-for-projects). I believe similar functionality exists for other languages like Python and JavaScript (Node.js). Just note that Homebrew's official tap only provides Long-Term Support (LTS) and the latest versions (for OpenJDK: 8, 11, 17, 21).

[^1]: On macOS, `openjdk` formulae are not in `PATH` due to the presence of the system Java wrappers at `/usr/bin/java`, `/usr/bin/javac`, etc. You can follow the caveats in the output of `brew info openjdk` to make them visible to the wrappers.

**Reply by RootRooster** [reply]
Thank you very much for your swift and insightful response. I truly appreciate the information provided. I will certainly look into the proposal you mentioned regarding multiple cellars in Homebrew.

Regarding the VS Code extension for configuring different Java runtimes, it sounds quite intriguing! However, I primarily utilize dev containers, which have been remarkably effective for managing multiple development environments seamlessly. If you haven't already, I highly recommend exploring the [devcontainers extension](https://code.visualstudio.com/docs/devcontainers/tutorial) for VS Code as it addresses most of my needs effortlessly.

While dev containers have been incredibly useful, there are still a few tasks that necessitate working in my personal development environment. For instance, certain tasks require access to my display, which Docker currently doesn't support. However, overall, the devcontainers extension has significantly streamlined my workflow.

Once again, thank you for your valuable input and suggestions. I look forward to exploring further and optimizing my development setup.
